command/init.go

```diff
@@ -627,14 +633,53 @@ func (c *InitCommand) getProviders(config *configs.Config, state *states.State,
 			}
 		}
 		if len(foundProviders) > 0 {
+			// Build list of provider suggestions, and track a list of local
+			// and remote modules which need to be upgraded
 			var providerSuggestions string
+			localModules := make(map[string]struct{})
+			remoteModules := make(map[*configs.ModuleRequirements]struct{})
 			for missingProvider, foundProvider := range foundProviders {
 				providerSuggestions += fmt.Sprintf("  %s -> %s\n", missingProvider.ForDisplay(), foundProvider.ForDisplay())
+				exists := struct{}{}
+				for _, reqs := range providerToReqs[missingProvider] {
+					src := reqs.SourceAddr
+					if src == "" || isLocalSourceAddr(src) {
+						localModules[reqs.SourceDir] = exists
+					} else {
+						remoteModules[reqs] = exists
+					}
+				}
+			}
+
+			// Create sorted list of 0.13upgrade commands with the affected
+			// source dirs
+			var upgradeCommands []string
+			for dir := range localModules {
+				upgradeCommands = append(upgradeCommands, fmt.Sprintf("terraform 0.13upgrade %s", dir))
+			}
+			sort.Strings(upgradeCommands)
+			command := "command"
+			if len(upgradeCommands) > 1 {
+				command = "commands"
+			}
+
+			// Display detailed diagnostic results, including the missing and
+			// found provider FQNs, and the suggested series of upgrade
+			// commands to fix this
+			detail := fmt.Sprintf("Could not find required providers, but found possible alternatives:\n\n%s\nIf these suggestions look correct, upgrade your configuration with the following %s:", providerSuggestions, command)
```

Comment:
I might've done this with [`strings.Builder`](https://godoc.org/strings#Builder) and then `fmt.Fprintf` rather than string concatenation, but in this context that's really just a subjective style thing and so I'm raising it just in case it wasn't something you considered; I'm fine with sticking with the string concatenation if you prefer it, or even if you're just neutral on it (since this codepath should only live for one major release anyway).

Reply:
Ah, that's clearer. Thanks for the suggestion, I hadn't thought of it!